GPII-17: Move gpii-oauth2 to gpii/universal/gpii/node_modules

simonbates opened the pull request.

amb26 commented on gpii/node_modules/flowManager/src/CloudBasedFlowManager.js:
```diff
@@ -0,0 +1,245 @@
+gpii.flowManager.cloudBased.privacy.filterOneApplication = function (application, selectedPreferences) {
+    var settings = application.settings;
+    var settingsInPrivacy = fluid.model.transformWithRules(settings, gpii.flowManager.cloudBased.privacy.forwardTransform);
+    console.log("GOT SETTINGS IN PRIVACY ONTOLOGY " + JSON.stringify(settingsInPrivacy, null, 2));
```
We should remove these console logs or replace them with calls to fluid.log before committing to master

simonbates replied:
Antranig, these console logs are your code -- which is your preference to remove or replace with fluid.log?

simonbates replied:
I will turn these into fluid.log

simonbates replied:
Done.

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/lib/foundation/js/foundation/foundation.dropdown.js:
```diff
@@ -0,0 +1,326 @@
+;(function ($, window, document, undefined) {
```
I don't believe this file is used in our application - we should remove it from the commit

jobara replied:
Fixed in https://github.com/simonbates/universal/pull/12

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/lib/foundation/js/foundation/foundation.js:
```diff
@@ -0,0 +1,625 @@
+/*
```
It doesn't seem that any Foundation JS is used in our application - we should remove it from the commit

jobara replied:
Fixed in simonbates#12

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/lib/infusion/src/lib/jquery/ui/css/fl-theme-bw/bw.css:
```diff
@@ -0,0 +1,1177 @@
+/*! jQuery UI - v1.10.3 - 2013-10-21
```
Could we avoid committing themes that are not actually used? Pending a solution for including material like this from npm

jobara replied:
Fixed in simonbates#12

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/core/js/GPIIOauth2.js:
```diff
@@ -0,0 +1,32 @@
+    fluid.registerNamespace("gpii.oauth2");
+
+    gpii.oauth2.ajax = function (urlTemplate, urlParams, options) {
+        var url = fluid.stringTemplate(urlTemplate, urlParams);
```
This URL interpolation does not apply proper escaping

jobara replied:
Fixed in simonbates#12

simonbates replied:
I have changed gpii.oauth2.ajax from the version in https://github.com/simonbates/universal/pull/12

https://github.com/simonbates/universal/commit/14164be56154b6711d84b90d58cb47672a5fd16e

The initial fix was to call encodeURI() on the result of fluid.stringTemplate(). This encoding will be too late for values which contain characters with special meanings in the URI query part such as "?" and "&" -- encodeURI() will leave them as is.

I have changed the code to instead call encodeURIComponent() on each individual replacement value before insertion into the template.

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/privacySettings/js/PrivacySettingsWithPrefs.js:
```diff
@@ -0,0 +1,240 @@
+        var dialogContent = fluid.stringTemplate(that.options.strings.removeDecisionContent, {serviceName: clientData.serviceName});
+        dialogForRemoval.find(that.options.selectors.removeDecisionContent).html(dialogContent);
+
+        var dialogOptions = {
```
Please refactor all of this inline material into a basic dialog component. This can for now have a single concrete grade OKCancelDialog with the two buttons hardwired and events named clickCancel and clickOK. If you are feeling more ambitious you can turn this into a model-based component with an area initially looking like 
clicks: {
    okButton: 0,
    cancelButton: 0
} (synthesized from the "buttons" configuration) and then forward every "click" callback from the button onto a listener which adds 1 to the relevant click count. All of the external click listeners can then be implemented as modelListeners rather than plain listeners.

jobara replied:
@simonbates could you please take a look into this. I don't think @cindyli and I will have time to make this change.

simonbates replied:
I've extracted a new OKCancelDialog component in:

https://github.com/simonbates/universal/commit/70e7457e7fc3c61ce399f2d5870366969885ce9b

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/tests/components/privacySettings/js/PrivacySettingsWithPrefsTests.js:
```diff
@@ -0,0 +1,51 @@
+            }
+        });
+
+        var assertRenderedText = function (that, root, paths, method) {
```
Add these to a global namespace and register them as strings rather than function literals

jobara replied:
Fixed in simonbates#12

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/tests/components/selectionTree/js/SelectionTreeTests.js:
```diff
@@ -0,0 +1,810 @@
+            "c.k.m": true
+        };
+
+        gpii.tests.oauth2.selectionTree.sampleComponetModel = {
```
Typo here Component

jobara replied:
Fixed in simonbates#12

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/src/AuthServer.js:
```diff
@@ -0,0 +1,511 @@
+
+    // TODO the /gpii_token endpoint was added to integrate Easit4All for the January 2015 review
+    // TODO it should be removed after the review GPII-1069
+    that.expressApp.get("/gpii_token",
```
Can we remove this endpoint now?

simonbates replied:
Done.

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/views/authorize.handlebars:
```diff
@@ -0,0 +1,51 @@
+                user: "{{user.username}}",
+                service: "{{client.name}}",
+                clientID: "{{client.oauth2ClientId}}",
+                transactionID: "{{ transactionID }}"
```
Are the spaces in this interpolation significant?

jobara replied:
Fixed in simonbates#12

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/tests/components/selectionTree/js/SelectionTreeTests.js:
```diff
@@ -0,0 +1,810 @@
+
+        jqUnit.test("gpii.oauth2.selectionTree.removeLeaf", function () {
+            gpii.tests.oauth2.selectionTree.testRemoveLeaf("Top Level", ".gpiic-ouath2-selectionTree-removeLeaf-topLevel");
+            gpii.tests.oauth2.selectionTree.testRemoveLeaf("Nested", ".gpiic-ouath2-selectionTree-removeLeaf-nested");
```
typo here ouath2

jobara replied:
Fixed in simonbates#12

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/tests/components/selectionTree/html/SelectionTree-test.html:
```diff
@@ -0,0 +1,97 @@
+            </ul>
+        </div>
+
+        <div class="gpiic-ouath2-selectionTree-removeLeaf">
```
typos throughout here ouath2

jobara replied:
Fixed in simonbates#12

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/selectionTree/js/SelectionTree.js:
```diff
@@ -0,0 +1,442 @@
+                        if (path) {
+                            paths.push(seg + "." + path);
+                        } else {
+                            paths.push(seg);
```
This method of terminating the recursion is pretty irregular. Factorise this into two functions, a top-level driver which accepts model, dereferences value and sets up path, and a separate function capturing the recursion. The recursion termination should occur at an outer level of this function (hopefully simply as a byproduct of the type check - clearer to use negation of isPrimitive rather than isPlainObject) rather than so deeply nested

jobara replied:
@simonbates could you please take a look into this. I don't think @cindyli and I will have time to make this change.

simonbates replied:
I've refactored gpii.oauth2.selectionTree.gatherPaths and I think the logic is a little clearer now:

https://github.com/simonbates/universal/commit/5f2b13f6f892363bcf40893a74ddecb656a2ce83

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/selectionTree/js/SelectionTree.js:
```diff
@@ -0,0 +1,442 @@
+        var collapseState;
+        var expandState;
+
+        if (arguments.length === 3) {
```
I would prefer a check against state === undefined to be less framework-hostile : P

jobara replied:
Fixed in simonbates#12

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/selectionTree/js/SelectionTree.js:
```diff
@@ -0,0 +1,442 @@
+                funcName:"gpii.oauth2.selectionTree.toServerModel",
+                args: ["{that}.model"]
+            },
+            updateModelFromServer: {
```
Add a TODO here that the whole "serverModel" logic in this component needs to be offloaded from this implementation, as we discussed in December. We can leave this task for when we have more time and are refactoring this whole UI to be payload- rather than markup-driven

jobara replied:
Fixed in simonbates#12

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/privacySettings/js/EditPrivacySettings.js:
```diff
@@ -0,0 +1,241 @@
+                funcName: "gpii.oauth2.editPrivacySettings.savePrefsAndExit",
+                args: ["{that}"]
+            },
+            setDoneState: {
```
This should most likely go into the model and the UI action implemented as a listener

jobara replied:
@simonbates could you please take a look into this. I don't think @cindyli and I will have time to make this change.

simonbates replied:
I've filed a JIRA issue for this work:

http://issues.gpii.net/browse/GPII-1124

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/privacySettings/js/EditPrivacySettings.js:
```diff
@@ -0,0 +1,241 @@
+                args: ["{that}.options.requestInfos.fetchAvailableAuthorizedPrefs.url", {
+                    clientID: "{that}.model.clientData.oauth2ClientId"
+                }, {
+                    //TODO: Handle errors
```
At least log them!

simonbates replied:
I've added a default error handler to gpii.oauth2.ajax:

https://github.com/simonbates/universal/commit/768ae7ed9481e537232378afc0704b7e2f61cc93

jobara commented on examples/README.md:
```diff
@@ -0,0 +1,53 @@
+
+- [http://localhost:3002/](http://localhost:3002/) - Sample client
+- [http://localhost:3003/](http://localhost:3003/) - Sample client with Passport
+- [http://localhost:3000/privacy](http://localhost:3000/privacy) - Privacy settings
```
I believe the port is now 8081, this should be updated to reflect that.

simonbates replied:
Done: https://github.com/simonbates/universal/commit/9695b7dae16af1db38ef56e602314a5238fd97da

jobara commented:
@simonbates there seems to be a bug with editing the privacy settings. Can you verify this.

Steps to reproduce:

1) Launch the Oauth2 single server and the sample-client-passport

2) From the sample client, request authorization

3) Login in with Alice and accept the authorization for increase-size

4) go to the privacy settings page

5) Edit the authorization ( the selection should be correct ), trigger the "done" button.

6) Re-open the edit dialog. Notice that all of the selections have been removed.

simonbates commented:
I have addressed the privacy settings editing issue that Justin discovered:

https://github.com/simonbates/universal/commit/9695b7dae16af1db38ef56e602314a5238fd97da

It was caused by the gpii.oauth2.authServer.standalone grade missing a JSON body-parser. We didn't see this issue earlier as once we integrated with the Cloud Based Flow Manager, we started running that and no longer used the standalone version of the AuthServer. I have fixed it for now, but in the future, if we no longer have use for the standalone server, we should just remove it so that we don't spend time maintaining something that is unnecessary.

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/selectionTree/js/SelectionTree.js:
```diff
@@ -0,0 +1,445 @@
+                fluid.each(model, function (subModel, seg) {
+                    var subPaths = gpii.oauth2.selectionTree.gatherPaths(subModel);
+                    fluid.each(subPaths, function (subPath) {
+                        if (subPath) {
```
In what situation might subPath be falsy - is this related to the paths.push("") above?
I think this logic would be clearer if the current model prefix and the total path list were passed as arguments to the recursion, rather than the return values from each recursion needing to be reprocessed by prefixing. We should arrange that each path element is added to "paths" just once only and with the correct full path at the point of addition.

amb26 commented on gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/selectionTree/js/SelectionTree.js:
```diff
@@ -0,0 +1,443 @@
+    };
+
+    gpii.oauth2.selectionTree.toModel = function (setPrefs, clientAvailablePrefs) {
+        var model = {};
```
Thanks, Simon - this implementation is looking good now. Could we here, and in all the called utility functions setValueAndXxxx, rename "model" to "newModel" to emphasise that, in for example, setValueAndAllAncestors, we are not modifying the component's model directly but only an isolated in-memory structure that we will submit later via the ChangeApplier.

simonbates replied:
Done in https://github.com/simonbates/universal/commit/b757f6f1de0bcd1b491aee2093c2626b45b5e7d2

Though I used "newSelections" rather than "newModel" as I moved the selections tree in the model to a container called "selections" -- to accommodate the new "hasSelection" boolean value. model.selections and model.hasSelection are siblings. See https://github.com/simonbates/universal/blob/b757f6f1de0bcd1b491aee2093c2626b45b5e7d2/gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/selectionTree/js/SelectionTree.js#L42-L60
